src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -0,0 +1,823 @@
+                     expectedValues:(NSArray<NSDictionary<NSString *, id> *> *)expectedValues
+              expectedValueInterval:(NSNumber *)expectedValueIntervalMs
+                 timedInvokeTimeout:(NSNumber * _Nullable)timeoutMs
+                        clientQueue:(dispatch_queue_t)clientQueue
+                         completion:(MTRDeviceResponseHandler)completion
+{
+    // Perform this operation
+    [self connectAndPerformAsync:^(MTRBaseDevice * baseDevice) {
+        [baseDevice
+            invokeCommandWithEndpointId:endpointId
+                              clusterId:clusterId
+                              commandId:commandId
+                          commandFields:commandFields
+                     timedInvokeTimeout:timeoutMs
+                            clientQueue:self.queue
+                             completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values, NSError * _Nullable error) {
+                                 dispatch_async(clientQueue, ^{
+                                     completion(values, error);
+                                 });
+                             }];
+    }];
+
+    [self setExpectedValues:expectedValues expectedValueInterval:expectedValueIntervalMs];
+}
+
+#pragma mark - Cache management
+
+// (assume lock is held)
+- (void)_checkExpiredExpectedValues
+{
+}
+
+// Get attribute value dictionary for an attribute path from the right cache
+- (NSDictionary<NSString *, id> *)_attributeValueDictionaryForAttributePath:(MTRAttributePath *)attributePath
+{
+    os_unfair_lock_lock(&self->_lock);
+    // First check expected value cache
+    MTRPair<NSDate *, NSDictionary *> * expectedValue = _expectedValueCache[attributePath];
+    if (expectedValue) {
+        NSDate * now = [NSDate date];
+        if ([now compare:expectedValue.first] == NSOrderedDescending) {
+            // expired - purge and fall through
+            _expectedValueCache[attributePath] = nil;
+        } else {
+            // not yet expired - return result
+            return expectedValue.second;
+        }
+    }
+
+    // Then check read cache
+    NSDictionary<NSString *, id> * cachedAttributeValue = _readCache[attributePath];
+    if (cachedAttributeValue) {
+        return cachedAttributeValue;
+    } else {
+        // TODO: when not found in cache, generated default values should be used
+        MTR_LOG_INFO(
+            "_attributeValueDictionaryForAttributePath: could not find cached attribute values for attribute %@", attributePath);
+    }
+
+    return nil;
+}
+
+- (BOOL)_attributeDataValue:(NSDictionary *)one isEqualToDataValue:(NSDictionary *)theOther
+{
+    // Attribute data-value dictionary should be all standard containers which
+    // means isEqual: comparisons all the way down, making a deep comparison.
+    return [one isEqualToDictionary:theOther];
+}
+
+- (NSArray *)_getAttributesToReportWithReportedValues:(NSArray<NSDictionary<NSString *, id> *> *)reportedAttributeValues
```

Comment:
This methof would benefit from some documentation describing how the return value relates to the incoming data and the caches.